core/src/main/java/com/linecorp/armeria/server/docs/JsonSchemaGenerator.java

```diff
@@ -135,61 +191,66 @@ private ObjectNode generate(MethodInfo methodInfo) {
      * and add required fields to the {@link ArrayNode}.
      *
      * @param field field to generate schema for
-     * @param visited map of visited types and their paths
      * @param path current path in tree traversal of fields
      * @param parent the parent to add schema properties
      * @param required the array node to add required field names, if parent doesn't support, it is null.
      */
-    private void generateField(FieldInfo field, Map<TypeSignature, String> visited, String path,
-                               ObjectNode parent,
-                               @Nullable ArrayNode required) {
+    private void generateField(FieldInfo field, String path, ObjectNode parent, @Nullable ArrayNode required) {
         final ObjectNode fieldNode = mapper.createObjectNode();
         final TypeSignature fieldTypeSignature = field.typeSignature();
+        final String schemaType = getSchemaType(fieldTypeSignature);
+        final String currentPath;
 
-        fieldNode.put("description", field.descriptionInfo().docString());
+        if (field.name().isEmpty()) {
+            currentPath = path;
+        } else {
+            currentPath = path + '/' + field.name();
+        }
+
+        if (!field.descriptionInfo().docString().isEmpty()) {
+            fieldNode.put("description", field.descriptionInfo().docString());
+        }
 
         // Fill required fields for the current object.
         if (required != null && field.requirement() == FieldRequirement.REQUIRED) {
             required.add(field.name());
         }
 
+        final boolean shouldFlatten = useFlatSchema &&
+                                      MEMORIZED_JSON_TYPES.contains(schemaType) &&
+                                      currentPath.chars().filter(c -> c == '/').count() > 1;
+        if (shouldFlatten) {
+            // If field is already visited, add a reference to the field instead of iterating its children.
+            if (!visited.containsKey(fieldTypeSignature)) {
+                final ObjectNode defsFieldNode = mapper.createObjectNode();
+                final String currentDefsPath = DEFS_PATH + '/' + fieldTypeSignature.signature();
+                visited.put(fieldTypeSignature, currentDefsPath);
+
+                generateTypeFields(defsFieldNode, field, schemaType, currentDefsPath);
+
+                defsNode.set(fieldTypeSignature.signature(), defsFieldNode);
+            }
+        }
+
         if (visited.containsKey(fieldTypeSignature)) {
             // If field is already visited, add a reference to the field instead of iterating its children.
             final String pathName = visited.get(fieldTypeSignature);
             fieldNode.put("$ref", pathName);
```

Comment:
So now, it seems like we add `pathName` which might contain a def with `$ref`.
However, is it not allowed to do that because of the recursion?
```
however, that a $ref referring to another $ref could cause an infinite loop in the resolver, and is explicitly disallowed.
```
https://json-schema.org/understanding-json-schema/structuring.html#id20

Reply:
Ah good catch, let me think about this for some time.

Reply:
I am still not sure how to implement this without using recursion. The explanation did not make sense to me. As there can be recursive definitions such as a family tree structure, a $ref should eventually reference itself or another ref.

Reply:
This is a shower thought,

I think what they mean is the following in the example:

```json
{
  "$defs": {
    "alice": { "$ref": "#/$defs/bob" },
    "bob": { "$ref": "#/$defs/alice" }
  }
}
```

In this case a ref refers to another ref. It doesn't mean it can't refer to another ref that has a ref. The rule is that body can't be a ref. I.e. following would be allowed:

```json
{
  "$defs": {
    "alice": { "$ref": "#/$defs/bob" },
    "bob": { "properties": {"foo": { "$ref": "#/$defs/notAlice" } } } 
  }
}
```

alice can ref to bob, but bob can't ref to alice, because alice is basically a pointer, it doesn't have any info.

Does that make sense? @minwoox 😆

Reply:
> The rule is that body can't be a ref. I.e. following would be allowed:

What if the schema is something like:
```
{
  "$defs": {
    "alice": { "$ref": "#/$defs/bob" },
    "bob": { "properties": {"foo": { "$ref": "#/$defs/alice" } } } 
  }
}
```

> It doesn't mean it can't refer to another ref that has a ref.

I'm not sure about this. The child property of the referenced property might have a reference to the referencing property which causes recursion.
How about allowing referencing to a property that doesn't have any reference in its hierarchy or only has a reference to itself?

Reply:
> I'm not sure about this. The child property of the referenced property might have a reference to the referencing property which causes recursion.

This should be fine in my opinion. Let me try to prove it because validation should be done lazily. For example,

Let's use https://www.jsonschemavalidator.net to validate our schemas,
```json
{
  "properties": {
    "foo": {
      "$ref": "#/$defs/alice"
    }
  },
  "$defs": {
    "alice": {
      "$ref": "#/$defs/bob"
    },
    "bob": {
      "$ref": "#/$defs/alice"
    }
  }
}
```

this is not working because it can't lazily evaluate a `ref` referring to another `ref`. But if we refer to a `ref` with a `properties` block such as,

```Json
{
  "properties": {
    "foo": {
      "$ref": "#/$defs/alice",
    }
  },
  "additionalProperties": false,
  "$defs": {
    "alice": {
      "$ref": "#/$defs/bob",
    },
    "bob": {
      "type": "object",
      "properties": {
        "bar": {
          "type": "object",
          "$ref": "#/$defs/alice"
        }
      },
      "additionalProperties": false
    }
  }
}
```

it is working. So, does this make sense in that context?

Reply:
> How about allowing referencing to a property that doesn't have any reference in its hierarchy or only has a reference to itself?

This is roughly the current approach, so it won't make sense to flatten schemas that way 😆

Reply:
> But if we refer to a ref with a properties block such as,

I didn't know that it was acceptable. Then, let's do this in this way. Thanks! 😆